pkg/controller/plan/migrator/base/migrator.go

```diff
@@ -0,0 +1,318 @@
+	if r.Context.Plan.Spec.Warm {
+		itinerary = WarmItinerary
+	} else {
+		itinerary = ColdItinerary
+	}
+	itinerary.Predicate = predicate
+	return
+}
+
+func (r *BaseMigrator) Next(status *plan.VMStatus) (next string) {
+	itinerary := r.Itinerary(&BasePredicate{vm: &status.VM, context: r.Context})
+	step, done, err := itinerary.Next(status.Phase)
+	if done || err != nil {
+		next = Completed
+		if err != nil {
+			log.Error(err, "Next phase failed.")
+		}
+	} else {
+		next = step.Name
+	}
+	r.Log.Info("Itinerary transition", "current phase", status.Phase, "next phase", next)
+	return
+}
+
+func (r *BaseMigrator) ExecutePhase(vm *plan.VMStatus) (ok bool, err error) {
+	// return ok = false to delegate to default itinerary implementation
+	// in plan/migration.go
+	return
+}
+
+// Step gets the name of the pipeline step corresponding to the current VM phase.
+func (r *BaseMigrator) Step(status *plan.VMStatus) (step string) {
+	switch status.Phase {
+	case Started, CreateInitialSnapshot, WaitForInitialSnapshot, StoreInitialSnapshotDeltas, CreateDataVolumes:
+		step = Initialize
+	case AllocateDisks:
+		step = DiskAllocation
+	case CopyDisks, CopyingPaused, RemovePreviousSnapshot, WaitForPreviousSnapshotRemoval, CreateSnapshot, WaitForSnapshot, StoreSnapshotDeltas, AddCheckpoint, ConvertOpenstackSnapshot, WaitForDataVolumesStatus:
+		step = DiskTransfer
+	case RemovePenultimateSnapshot, WaitForPenultimateSnapshotRemoval, CreateFinalSnapshot, WaitForFinalSnapshot, AddFinalCheckpoint, Finalize, RemoveFinalSnapshot, WaitForFinalSnapshotRemoval, WaitForFinalDataVolumesStatus:
+		step = Cutover
+	case CreateGuestConversionPod, ConvertGuest:
+		step = ImageConversion
+	case CopyDisksVirtV2V:
+		step = DiskTransferV2v
+	case CreateVM:
+		step = VMCreation
+	case PreHook, PostHook:
+		step = status.Phase
+	case StorePowerState, PowerOffSource, WaitForPowerOff:
+		if r.Context.Plan.Spec.Warm {
+			step = Cutover
+		} else {
+			step = Initialize
+		}
+	default:
+		step = Unknown
+	}
+	return
+}
+
+// Step predicate.
+type BasePredicate struct {
+	// VM listed on the plan.
+	vm *plan.VM
+	// Plan context
+	context *plancontext.Context
+}
+
+// Evaluate predicate flags.
+func (r *BasePredicate) Evaluate(flag libitr.Flag) (allowed bool, err error) {
+	useV2vForTransfer, vErr := r.context.Plan.ShouldUseV2vForTransfer()
+	if vErr != nil {
+		err = vErr
+		return
+	}
+
+	switch flag {
+	case HasPreHook:
+		_, allowed = r.vm.FindHook(PreHook)
+	case HasPostHook:
+		_, allowed = r.vm.FindHook(PostHook)
+	case RequiresConversion:
+		allowed = r.context.Source.Provider.RequiresConversion()
+	case CDIDiskCopy:
+		allowed = !useV2vForTransfer
+	case VirtV2vDiskCopy:
+		allowed = useV2vForTransfer
+	case OpenstackImageMigration:
+		allowed = r.context.Plan.IsSourceProviderOpenstack()
+	case VSphere:
+		allowed = r.context.Plan.IsSourceProviderVSphere()
+	}
+
+	return
+}
+
+func (r *BasePredicate) Count() int {
```

Comment:
What's the usage of this?

Reply:
In the current system, there is one migration predicate and there is just a constant with a hard coded number of flags in the pkg/lib/itinerary package. The purpose of this is to encapsulate it behind the Predicate interface. It will make slightly more sense once I add a file that got missed when I committed...